library/std/src/os/uefi/env.rs

```diff
@@ -0,0 +1,54 @@
+//! UEFI-specific extensions to the primitives in `std::env` module
+
+use crate::ffi::c_void;
+use crate::ptr::NonNull;
+use crate::sync::atomic::{AtomicPtr, Ordering};
+use crate::sync::OnceLock;
+
+// Position 0 = SystemTable
+// Position 1 = ImageHandle
+static GLOBALS: OnceLock<(AtomicPtr<c_void>, AtomicPtr<c_void>)> = OnceLock::new();
```

Comment:
I see; I thought that AtomicPtr would prevent automatic pointer caching (manual is up to the user if they really know what they are doing). I think it should be fine to convert it into a combination of `NonNull` + `read_volatile`.

For more context, the SystemTable pointer can be modified from other programs (or UEFI events are async), Eg: in case of a device being plugged in or removed. How that is handled is up to the UEFI implementation but it is generally not considered a good idea to cache the SystemTable pointer.

Reply:
Okay, so considering our `*mut SystemTable`... just to be excruciatingly clear, can the SystemTable's position, itself, be modified? In that case, how would we possibly update on that fact? Some other predefined rendezvous point?

Or do you mean the "fields" of the SystemTable that we are pointing to?

Reply:
As far as I know, the SystemTable's position should not move, even after `ExitBootServices()` is called.
However, the pointers in SystemTable may become valid/invalid during the UEFI lifecycle.

Eg: After an operating system has taken control of the platform with a call to ExitBootServices(), only the Hdr, FirmwareVendor, FirmwareRevision, RuntimeServices, NumberOfTableEntries, and ConfigurationTable fields are valid.

Reply:
Okay.
Note:

> Just like in C, whether an operation is volatile has no bearing whatsoever on questions involving concurrent access from multiple threads. Volatile accesses behave exactly like non-atomic accesses in that regard. In particular, a race between a `read_volatile` and any write operation to the same location is undefined behavior.

And... I could be mistaken, I will doublecheck (I know the hardware semantics better than the compiler semantics, here), but loading a pointer from an AtomicPtr and then reading what is behind it doesn't seem like it would necessarily guarantee the read is atomic/synchronizing. AtomicPtr defines the updates to the pointer value itself as synchronizing, and that may discourage some optimizations, but if the compiler can provably observe the pointer value didn't change? I believe it's allowed to start reasoning about duplicate reads again (i.e. to start "caching" once more).

It may be that the load orderings happen to make that impossible, but it seems like we're expressing the wrong thing, in that case.

Reply:
This is just a summary of what is expected when storing SystemTable Pointer:
1. The position of SystemTable should be fixed.
2. Fields can become valid/invalid throughout the lifetime of UEFI.
3. Asynchronous modification of SystemTable is possible (Using Events).
4. UEFI is single-threaded; thus, parallel modification of SystemTable is impossible.

Some general considerations when storing it:
1. We probably don't want to re-export `r-efi` from std.
2. Discourage SystemTable pointer caching.

Reply:
Well, I have converted the globals to thread_locals as `Cell<Option<(NonNull<c_void>, NonNull<c_void>)>>`. Feel free to start a new thread if that might have some problems.